src/Swift.Runtime/src/Library/Arc.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+/// <summary>
+/// Arc is a class containing p/invokes for Swift Automatic Reference Counting and memory management.
+/// </summary>
+public static class Arc {
+    const string swiftCore = "/usr/lib/swift/libswiftCore.dylib";
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport(swiftCore, CallingConvention = CallingConvention.Cdecl)]
+    static extern void swift_retain(IntPtr p);
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static IntPtr Retain(IntPtr p)
```

Comment:
The current API design guideline is to use unmanaged pointer type (e.g. `void*`) for unmanaged pointers. Would it be appropriate here?

Reply:
@jkotas - I'm curious about the guideline here. The pointer is opaque to us on the managed side (in fact if we ever look inside it would be a problem). It's more like a handle for the managed side. `void *` implies `unsafe` which doesn't seem to fit these cases. There are actually quite a few places where we need to store the Swift's reference to the unmanaged object - if all of those would need to be `unsafe` it would mean effectively everything we do would need to be `unsafe`.

I would prefer that for places where we treat it as a handle, we keep it IntPtr (or maybe even some our struct which wraps it). And the very few places where we might need to treat it as a pointer, those would need to be `unsafe` since we would need to start using unmanaged pointers.

Reply:
If it is more like a handle, IntPtr is fine.